### PR TITLE
Make `core.Tx` a concrete type and remove `DefineTx` and `PersistTx`

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#634 

# Overview

- [x] We now only support one backend node again. I have removed the `DefineTx` and `PersistTx`-classes, and added the Jörmungandr specific `Tx`-type  I have made the core `Tx` (with `resolvedInputs :: [(TxIn, Coin)]`) to `core`.
- [x] I have removed `t` type parameters wherever possible. This lead to some `@t` applications being required.
- [x] I removed some CBOR encoding/decoding we won't need (we need to decode addresses, but not blocks)

# Comments

### CODE REVIEW

File: lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs [hunk 1]
```diff
@@ -253,7 +253,7 @@ spec = do
             , expectEventually ctx getWalletEp balanceTotal (2*amt)
             ]
 
-    it "TRANS_CREATE_02 - Multiple Output Tx to different wallets" $ \ctx -> do
+    it "TRANS_CREATE_02 - Multiple OutputTxo different wallets" $ \ctx -> do
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Scenario/API/Transactions.hs]: 🤔 Need to look over this bad find and replace

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 2]
```diff
@@ -504,15 +501,14 @@ postWallet ctx body = do
     let s = mkSeqState (rootXPrv, pwd) g
     let wid = WalletId $ digest $ publicKey rootXPrv
     let wName = getApiT (body ^. #name)
-    void $ liftHandler $ createWallet ctx wid wName s
-    liftHandler $ withWorkerCtx ctx wid throwE $ \wrk ->
-        W.attachPrivateKey wrk wid (rootXPrv, pwd)
+    void $ liftHandler $ createWallet @_ @s @t @k ctx wid wName s
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:504]: 😕 Had to add type-applications here

**Comment by Anviking** [comment]
```
 test/unit/Cardano/Wallet/DB/Properties.hs:201:9:
  1) Cardano.Wallet.DB.MVar.MVar, sequential puts replace values in order, Tx History
       Assertion failed (after 10 tests and 11 shrinks):
         KeyValPairs [(948caa2d...95043914,- ( 30653037~> 1st 30623064
             <~ 16537 @ 2308503c...1627616d
           ,
             -0.000015 in ledger since 68.10#8 )
         )]
         (Wallet (UTxO {getUTxO = fromList [(TxIn {inputId = Hash {getHash = "00000f03"}, inputIx = 1},TxOut {address = Address {unAddress = "\"0F\t3~yj\ETB1\181\SI\163\STX\255,\b\240M\SOHN#E\205\ETX\EOT\171<&\221YZx"}, coin = Coin {getCoin = 68915}}),(TxIn {inputId = Hash {getHash = "0208020d"}, inputIx = 1},TxOut {address = Address {unAddress = "AM\\\SOwX#/%<-:\172\&7I7\DC1\222LA\b_(\215Q8\193!\ACKl\US\220M"}, coin = Coin {getCoin = 85769}})]}) (BlockHeader {slotId = SlotId {epochNumber = EpochNo {unEpochNo = 0}, slotNumber = SlotNo {unSlotNo = 0}}, blockHeight = Quantity {getQuantity = 0}, headerHash = Hash {getHash = "block0"}, parentHeaderHash = Hash {getHash = "genesis"}}) (SeqState {internalPool = AddressPool {accountPubKey = ShelleyKey {getKey = XPub {xpubPublicKey = "Q\170\GS\202\198\&2KA\203\CANN'X\154 \139\DEL\FS\148\FSb\SO\RS\r\DLEAL\151\153\137\167\194", xpubChaincode = ChainCode "\204\196\"I\225y\132\196L\243\128\180\137\246,W\248@\137\225P$[\244\156Cm\v\151\t\197\143"}}, gap = AddressPoolGap {getAddressPoolGap = 10}, indexedKeys = fromList [(KeyFingerprint "\v\149\SI\172\228\ESC\251C\254<\a\b:;\219\198\210\164\228I(\163\188\183\223\178\&3\241{~\165\154",Index {getIndex = 2}),(KeyFingerprint "@\255\173.\229\241\188\229\237\ng4\\I\242|\160\US\162\135\223w\140\138\201\NAKS\ETBR\149\197\186",Index {getIndex = 5}),(KeyFingerprint "F\216\165\232\&9\138\202\NAK\200\130\208l#\139\238\236\240\&7Q%\255\238I\NUL\238\181\"\EM\196\137\190\EM",Index {getIndex = 3}),(KeyFingerprint "`|\194\195\223\DC4\DC4a.\vJ\188\148\183\173Kd$\198\242\DEL\138=\143\141QQw\133gr%",Index {getIndex = 6}),(KeyFingerprint "\131\133\a\172;NK\DLE\191\194\168\252\239\DLE\221\\\216\RSM=\167\223\247\ETXX\214\239@\180\167\223\146",Index {getIndex = 0}),(KeyFingerprint "\147\189H\205\142\DEL\180\\:$\202\STXvF\157 *\GS\130[\134,\179\226C3!\EOT\228\r\n\a",Index {getIndex = 9}),(KeyFingerprint "\170\222\247\173\178\217_\208\a\222p\141\210\200\&4oG\180\210\214\239\&6\"|\183\SOH!\147\132/\146\235",Index {getIndex = 1}),(KeyFingerprint "\194+~\186\152\DC1\215\181\158\249\253\164u\175\EM\212\&7\US\SOH\174\247\139.*!\233\175\149\247\214\131\178",Index {getIndex = 8}),(KeyFingerprint "\244M\229\255k\249a\231\132\f\128\161\210\&9\209\191\&6a\SUB)n\141C\r\"\158\184B*\200\186\184",Index {getIndex = 4}),(KeyFingerprint "\252Nb\136\FS\CAN\v\222\FS\b\230M\247\225bT\235Y\201b\196\145\199\231\243\207\156|\209R\245\SYN",Index {getIndex = 7})]}, externalPool = AddressPool {accountPubKey = ShelleyKey {getKey = XPub {xpubPublicKey = "Q\170\GS\202\198\&2KA\203\CANN'X\154 \139\DEL\FS\148\FSb\SO\RS\r\DLEAL\151\153\137\167\194", xpubChaincode = ChainCode "\204\196\"I\225y\132\196L\243\128\180\137\246,W\248@\137\225P$[\244\156Cm\v\151\t\197\143"}}, gap = AddressPoolGap {getAddressPoolGap = 10}, indexedKeys = fromList [(KeyFingerprint "\r\225\208\DEL\196 \"m\144\140\250\187\147\158\163\189\ETB\216\fK(\185\ETX\168\144\206D\197\DC2\206I\RS",Index {getIndex = 2}),(KeyFingerprint "t\ESC\179\NUL\177\199j\241&>\175\201\212\215\GS,4g,\SOH\219\&8\223\186\ETBm\229\162\&2`\147\SOH",Index {getIndex = 8}),(KeyFingerprint "\145\233\140\ACK=G\146i\205\FSi\251\236\ETX*&\188\SYN\145\142\a` \151\155/\223\ENQ\ESC\233P\199",Index {getIndex = 4}),(KeyFingerprint "\148\246S\208\134\184\246\r\252\239\135\163o\r\ETB\162\176\134\131\bv\208\189W\SOH\176\165p\178\DLEg2",Index {getIndex = 0}),(KeyFingerprint "\172\190\"\ACK]\149l\"\232Nn\ETB\DC2\158\153\222(\SOH\ESCl\156\215\154\SOL\169\225T\247v\132\152",Index {getIndex = 3}),(KeyFingerprint "\203M\163c\EM\224%l\132\SOH\133\GS\195\253\ETB\DC3\172\191\241\180\158\EM\SI\USz\245\176\&7\EOT\238\193\216",Index {getIndex = 5}),(KeyFingerprint "\215\199\163\150zxb\156\159I}\t\220\171AY\220\223C\249@\SI\EOT\216\230\170\162\149\239\204\172\251",Index {getIndex = 1}),(KeyFingerprint "\221I:\DC2\174\&0`\218\ETB\203\225\243\234\&2\167G\SOH\241\DLE\vK\220q\n\213\129cFgjnp",Index {getIndex = 9}),(KeyFingerprint "\244\176Q\FS\FS\181K\183\214\ENQ=^u\235\204\213&\n;1!&305\ACK`fE\202\164<",Index {getIndex = 7}),(KeyFingerprint "\245\254\207\DC4<\167\142\&8\200\157%\161>\EOT@^\203\243\128\176\163bJ\166\174~\ETB2\146\&3u\US",Index {getIndex = 6})]}, pendingChangeIxs = PendingIxs {pendingIxsToList = []}}) (BlockchainParameters {getGenesisBlockHash = Hash {getHash = "genesis"}, getGenesisBlockDate = StartTime 1970-01-01 00:00:00 UTC, getFeePolicy = LinearFee (Quantity {getQuantity = 14.0}) (Quantity {getQuantity = 42.0}) (Quantity {getQuantity = 5.0}), getSlotLength = SlotLength 1s, getEpochLength = EpochLength 21600, getTxMaxSize = Quantity {getQuantity = 8192}, getEpochStability = Quantity {getQuantity = 2160}}),WalletMetadata {name = WalletName {getWalletName = "charmander"}, creationTime = 1938-11-23 12:30:17 UTC, passphraseInfo = Just (WalletPassphraseInfo {lastUpdatedAt = 2295-01-16 07:17:06 UTC}), delegation = NotDelegating})

         Resolved
         [- ( 30653037~> 1st 30623064
             <~ 16537 @ 2308503c...1627616d
           ,
             -0.000015 in ledger since 68.10#8 )
         ]

         Read
         [- ( 30653037~> 2nd 30363034
             <~ 89758 @ f87374bd...35253760
             <~ 2162 @ 855e4a21...41138f1f
             <~ 34984 @ 6f353bec...4d674042
             <~ 17959 @ 43415c26...4a21791c
             <~ 55918 @ 8120ff9b...06570128
           ,
             -0.000015 in ledger since 68.10#8 )
         ]
         Resolved == Read ✗

  To rerun use: --match "/Cardano.Wallet.DB.MVar/MVar/sequential puts replace values in order/Tx History/"

Randomized with seed 1294518536
```

**Comment by Anviking** [comment]
Fixed by b7e15a7, but computing hashes is probably what slows down the tests to timeout.

**Comment by rvl** [comment]
It probably won't be too long until we again have two backends to support.

So do you think the maintenance overhead saved by having a concrete Tx type will exceed the time spent removing the abstraction then later putting it back again?

**Comment by KtorZ** [comment]
@rvl The idea here is really to undo the abstraction to introduce another one later that'd be more satisfactory. Actually, with the Haskell nodes coming in, we'll likely have more than just the transaction type that is going to be different. So the little trick we did for the http-bridge vs jormungandr might not sufficient. And, in order to introduce a proper abstraction here (and probably avoid concrete types altogether in the core code) it's necessary to first remove what is now a wrong abstraction.

**Comment by KtorZ** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 3]
```diff
@@ -849,7 +832,7 @@ network ctx = do
 compatibilityApiServer
     :: forall t n.
         ( Buildable (ErrValidateSelection t)
-        , DefineTx t
+        , PaymentAddress n ByronKey
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: This addition seems unnecessary to me :thinking:

**Review by KtorZ — APPROVED** [review]

**Pull request merged**